
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="X-UA-Compatible" content="IE=Edge" />
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <title>Calculating summary statistics &#8212; pylibseq 0.2.2 documentation</title>
    <link rel="stylesheet" href="../_static/classic.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript" id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <script type="text/javascript" src="../_static/language_data.js"></script>
    <script async="async" type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
    
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Tree Sequences" href="treesequences.html" />
    <link rel="prev" title="The VariantMatrix" href="vmatrix.html" /> 
  </head><body>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="treesequences.html" title="Tree Sequences"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="vmatrix.html" title="The VariantMatrix"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="../index.html">pylibseq 0.2.2 documentation</a> &#187;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="calculating-summary-statistics">
<span id="summarrystats"></span><h1>Calculating summary statistics<a class="headerlink" href="#calculating-summary-statistics" title="Permalink to this headline">¶</a></h1>
<p>This page is a tutorial on calculating summary statistics from a <a class="reference internal" href="../index.html#libsequence.VariantMatrix" title="libsequence.VariantMatrix"><code class="xref py py-class docutils literal notranslate"><span class="pre">libsequence.VariantMatrix</span></code></a>.  We
also show how to interchange data between <a class="reference external" href="http://msprime.readthedocs.io">msprime</a> <a class="reference internal" href="zreferences.html#kelleher2016-cb" id="id1">[5]</a> and pylibseq.</p>
<div class="section" id="simple-overview-of-concepts">
<h2>Simple overview of concepts<a class="headerlink" href="#simple-overview-of-concepts" title="Permalink to this headline">¶</a></h2>
<div class="highlight-ipython notranslate"><div class="highlight"><pre><span></span><span class="gp">In [1]: </span><span class="kn">import</span> <span class="nn">msprime</span>

<span class="gp">In [2]: </span><span class="kn">import</span> <span class="nn">libsequence</span>

<span class="gp">In [3]: </span><span class="kn">import</span> <span class="nn">numpy</span> <span class="kn">as</span> <span class="nn">np</span>

<span class="go"># Get a TreeSequence from msprime</span>
<span class="gp">In [4]: </span><span class="n">ts</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">simulate</span><span class="p">(</span><span class="mi">100</span><span class="p">,</span> <span class="n">mutation_rate</span><span class="o">=</span><span class="mf">250.</span><span class="p">,</span> <span class="n">recombination_rate</span><span class="o">=</span><span class="mf">250.</span><span class="p">,</span> <span class="n">random_seed</span><span class="o">=</span><span class="mi">42</span><span class="p">)</span>

<span class="go"># TreeSequence -&gt; VariantMatrix</span>
<span class="gp">In [5]: </span><span class="n">vm</span> <span class="o">=</span> <span class="n">libsequence</span><span class="o">.</span><span class="n">VariantMatrix</span><span class="o">.</span><span class="n">from_TreeSequence</span><span class="p">(</span><span class="n">ts</span><span class="p">)</span>

<span class="go"># VariantMatrix -&gt; AlleleCountMatrix</span>
<span class="gp">In [6]: </span><span class="n">ac</span> <span class="o">=</span> <span class="n">vm</span><span class="o">.</span><span class="n">count_alleles</span><span class="p">()</span>
</pre></div>
</div>
<p>Certain summary statistics can be calculated simply from allele counts via
<a class="reference internal" href="../index.html#libsequence.AlleleCountMatrix" title="libsequence.AlleleCountMatrix"><code class="xref py py-class docutils literal notranslate"><span class="pre">libsequence.AlleleCountMatrix</span></code></a>. Calculations invovling linkage disequilibrium
will typically require the entire <a class="reference internal" href="../index.html#libsequence.VariantMatrix" title="libsequence.VariantMatrix"><code class="xref py py-class docutils literal notranslate"><span class="pre">libsequence.VariantMatrix</span></code></a> object.</p>
<p>The following examples rely on msprime <a class="reference internal" href="zreferences.html#kelleher2016-cb" id="id2">[5]</a> to generate input data, but
the concepts hold more generally.</p>
</div>
<div class="section" id="list-of-classic-standard-summary-statistics">
<h2>List of “classic”/standard summary statistics<a class="headerlink" href="#list-of-classic-standard-summary-statistics" title="Permalink to this headline">¶</a></h2>
<p>Nucleotide diversity, or <span class="math notranslate nohighlight">\(\pi\)</span> <a class="reference internal" href="zreferences.html#tajima1983-it" id="id3">[7]</a>:</p>
<dl class="function">
<dt id="libsequence.thetapi">
<code class="descclassname">libsequence.</code><code class="descname">thetapi</code><span class="sig-paren">(</span><em>ac: libsequence._libsequence.AlleleCountMatrix</em><span class="sig-paren">)</span> &#x2192; float<a class="headerlink" href="#libsequence.thetapi" title="Permalink to this definition">¶</a></dt>
<dd><p>Mean number of pairwise differences.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>ac</strong> – A <a class="reference internal" href="../index.html#libsequence.AlleleCountMatrix" title="libsequence.AlleleCountMatrix"><code class="xref py py-class docutils literal notranslate"><span class="pre">libsequence.AlleleCountMatrix</span></code></a></td>
</tr>
</tbody>
</table>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">Implemented as sum of site heterozygosity.</p>
</div>
</dd></dl>

<div class="highlight-ipython notranslate"><div class="highlight"><pre><span></span><span class="gp">In [7]: </span><span class="k">print</span><span class="p">(</span><span class="n">libsequence</span><span class="o">.</span><span class="n">thetapi</span><span class="p">(</span><span class="n">ac</span><span class="p">))</span>
<span class="go">980.3789898989783</span>
</pre></div>
</div>
<p>Watterson’s estimator of <span class="math notranslate nohighlight">\(\theta\)</span>, <a class="reference internal" href="zreferences.html#watterson1975-ej" id="id4">[10]</a>:</p>
<dl class="function">
<dt id="libsequence.thetaw">
<code class="descclassname">libsequence.</code><code class="descname">thetaw</code><span class="sig-paren">(</span><em>ac: libsequence._libsequence.AlleleCountMatrix</em><span class="sig-paren">)</span> &#x2192; float<a class="headerlink" href="#libsequence.thetaw" title="Permalink to this definition">¶</a></dt>
<dd><p>Watterson’s theta.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>m</strong> – A <a class="reference internal" href="../index.html#libsequence.AlleleCountMatrix" title="libsequence.AlleleCountMatrix"><code class="xref py py-class docutils literal notranslate"><span class="pre">libsequence.AlleleCountMatrix</span></code></a></td>
</tr>
</tbody>
</table>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">Calculated from the total number of mutations.</p>
</div>
</dd></dl>

<div class="highlight-ipython notranslate"><div class="highlight"><pre><span></span><span class="gp">In [8]: </span><span class="k">print</span><span class="p">(</span><span class="n">libsequence</span><span class="o">.</span><span class="n">thetaw</span><span class="p">(</span><span class="n">ac</span><span class="p">))</span>
<span class="go">982.5437651915445</span>
</pre></div>
</div>
<p>Tajima’s <span class="math notranslate nohighlight">\(D\)</span>, <a class="reference internal" href="zreferences.html#tajima1989-de" id="id5">[8]</a>:</p>
<dl class="function">
<dt id="libsequence.tajd">
<code class="descclassname">libsequence.</code><code class="descname">tajd</code><span class="sig-paren">(</span><em>ac: libsequence._libsequence.AlleleCountMatrix</em><span class="sig-paren">)</span> &#x2192; float<a class="headerlink" href="#libsequence.tajd" title="Permalink to this definition">¶</a></dt>
<dd><p>Tajima’s D.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>m</strong> – A <a class="reference internal" href="../index.html#libsequence.AlleleCountMatrix" title="libsequence.AlleleCountMatrix"><code class="xref py py-class docutils literal notranslate"><span class="pre">libsequence.AlleleCountMatrix</span></code></a></td>
</tr>
</tbody>
</table>
</dd></dl>

<div class="highlight-ipython notranslate"><div class="highlight"><pre><span></span><span class="gp">In [9]: </span><span class="k">print</span><span class="p">(</span><span class="n">libsequence</span><span class="o">.</span><span class="n">tajd</span><span class="p">(</span><span class="n">ac</span><span class="p">))</span>
<span class="go">-0.007527225314052745</span>
</pre></div>
</div>
<p>Fay and Wu’s test statistic, <span class="math notranslate nohighlight">\(\pi - \hat\theta_H\)</span> <a class="reference internal" href="zreferences.html#fay2000-ef" id="id6">[2]</a> has two overloads.
The first takes
a single ancestral state as an argument.  The latter takes a
list of ancestral states (one per site in the variant matrix):</p>
<dl class="function">
<dt id="libsequence.faywuh">
<code class="descclassname">libsequence.</code><code class="descname">faywuh</code><span class="sig-paren">(</span><em>*args</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#libsequence.faywuh" title="Permalink to this definition">¶</a></dt>
<dd><p>Overloaded function.</p>
<ol class="arabic simple">
<li>faywuh(ac: libsequence._libsequence.AlleleCountMatrix, ancestral_state: int) -&gt; float</li>
<li>faywuh(ac: libsequence._libsequence.AlleleCountMatrix, ancestral_states: List[int]) -&gt; float</li>
</ol>
</dd></dl>

<div class="highlight-ipython notranslate"><div class="highlight"><pre><span></span><span class="gp">In [10]: </span><span class="k">print</span><span class="p">(</span><span class="n">libsequence</span><span class="o">.</span><span class="n">faywuh</span><span class="p">(</span><span class="n">ac</span><span class="p">,</span> <span class="mi">0</span><span class="p">))</span>
<span class="go">-74.02989898988403</span>
</pre></div>
</div>
<p>The <span class="math notranslate nohighlight">\(H'\)</span> statistic <a class="reference internal" href="zreferences.html#zeng2006-is" id="id7">[11]</a> should be preferred to
the Fay and Wu test statistic:</p>
<dl class="function">
<dt id="libsequence.hprime">
<code class="descclassname">libsequence.</code><code class="descname">hprime</code><span class="sig-paren">(</span><em>*args</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#libsequence.hprime" title="Permalink to this definition">¶</a></dt>
<dd><p>Overloaded function.</p>
<ol class="arabic simple">
<li>hprime(ac: libsequence._libsequence.AlleleCountMatrix, ancestral_state: int) -&gt; float</li>
<li>hprime(ac: libsequence._libsequence.AlleleCountMatrix, ancestral_state: List[int]) -&gt; float</li>
</ol>
</dd></dl>

<div class="highlight-ipython notranslate"><div class="highlight"><pre><span></span><span class="gp">In [11]: </span><span class="k">print</span><span class="p">(</span><span class="n">libsequence</span><span class="o">.</span><span class="n">hprime</span><span class="p">(</span><span class="n">ac</span><span class="p">,</span> <span class="mi">0</span><span class="p">))</span>
<span class="go">-0.12962968409194914</span>
</pre></div>
</div>
<p>Summaries of haplotype variation <a class="reference internal" href="zreferences.html#depaulis1998-ol" id="id8">[1]</a>:</p>
<dl class="function">
<dt id="libsequence.number_of_haplotypes">
<code class="descclassname">libsequence.</code><code class="descname">number_of_haplotypes</code><span class="sig-paren">(</span><em>arg0: libsequence._libsequence.VariantMatrix</em><span class="sig-paren">)</span> &#x2192; int<a class="headerlink" href="#libsequence.number_of_haplotypes" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="function">
<dt id="libsequence.haplotype_diversity">
<code class="descclassname">libsequence.</code><code class="descname">haplotype_diversity</code><span class="sig-paren">(</span><em>arg0: libsequence._libsequence.VariantMatrix</em><span class="sig-paren">)</span> &#x2192; float<a class="headerlink" href="#libsequence.haplotype_diversity" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<div class="highlight-ipython notranslate"><div class="highlight"><pre><span></span><span class="gp">In [12]: </span><span class="k">print</span><span class="p">(</span><span class="n">libsequence</span><span class="o">.</span><span class="n">number_of_haplotypes</span><span class="p">(</span><span class="n">vm</span><span class="p">))</span>
<span class="go">99</span>

<span class="gp">In [13]: </span><span class="k">print</span><span class="p">(</span><span class="n">libsequence</span><span class="o">.</span><span class="n">haplotype_diversity</span><span class="p">(</span><span class="n">vm</span><span class="p">))</span>
<span class="go">   0.9997979797979798</span>
</pre></div>
</div>
<p>The <span class="math notranslate nohighlight">\(nS_L\)</span> and <span class="math notranslate nohighlight">\(iHS\)</span> “haplotype homozygosity” statistics
defined in <a class="reference internal" href="zreferences.html#ferrer-admetlla2014-wa" id="id9">[3]</a>:</p>
<dl class="function">
<dt id="libsequence.nsl">
<code class="descclassname">libsequence.</code><code class="descname">nsl</code><span class="sig-paren">(</span><em>m: libsequence._libsequence.VariantMatrix</em>, <em>refstate: int</em><span class="sig-paren">)</span> &#x2192; libsequence._libsequence.VecnSLResults<a class="headerlink" href="#libsequence.nsl" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<p>This function returns a list of instances of the following:</p>
<dl class="class">
<dt id="libsequence.nSLresults">
<em class="property">class </em><code class="descclassname">libsequence.</code><code class="descname">nSLresults</code><a class="headerlink" href="#libsequence.nSLresults" title="Permalink to this definition">¶</a></dt>
<dd><p>Holds nSL, iHS, and non-reference count at core SNP</p>
<dl class="attribute">
<dt id="libsequence.nSLresults.core_count">
<code class="descname">core_count</code><a class="headerlink" href="#libsequence.nSLresults.core_count" title="Permalink to this definition">¶</a></dt>
<dd><p>Core mutation count in sample</p>
</dd></dl>

<dl class="attribute">
<dt id="libsequence.nSLresults.ihs">
<code class="descname">ihs</code><a class="headerlink" href="#libsequence.nSLresults.ihs" title="Permalink to this definition">¶</a></dt>
<dd><p>iHS</p>
</dd></dl>

<dl class="attribute">
<dt id="libsequence.nSLresults.nsl">
<code class="descname">nsl</code><a class="headerlink" href="#libsequence.nSLresults.nsl" title="Permalink to this definition">¶</a></dt>
<dd><p>nSL</p>
</dd></dl>

</dd></dl>

<p>Values of <cite>nan</cite> imply that haplotype homozygosity for a particular core
variant was not broken within the matrix.  This is the same procedure
used in <a class="reference internal" href="zreferences.html#ferrer-admetlla2014-wa" id="id10">[3]</a>.  The return value order
corresponds to the row order of the input data:</p>
<div class="highlight-ipython notranslate"><div class="highlight"><pre><span></span><span class="gp">In [14]: </span><span class="n">nsl</span> <span class="o">=</span> <span class="n">libsequence</span><span class="o">.</span><span class="n">nsl</span><span class="p">(</span><span class="n">vm</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span>

<span class="gp">In [15]: </span><span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">nsl</span><span class="p">[:</span><span class="mi">4</span><span class="p">]:</span>
<span class="gp">   ....: </span>   <span class="k">print</span><span class="p">(</span><span class="n">i</span><span class="o">.</span><span class="n">core_count</span><span class="p">,</span> <span class="n">i</span><span class="o">.</span><span class="n">ihs</span><span class="p">,</span> <span class="n">i</span><span class="o">.</span><span class="n">nsl</span><span class="p">)</span>
<span class="gp">   ....: </span>
<span class="go">1 nan nan</span>
<span class="go">8 nan nan</span>
<span class="go">3 nan nan</span>
<span class="go">7 -0.09735588583687971 -0.5544227795580867</span>
</pre></div>
</div>
<p>The following variant only allows haplotype homozygosity to be
broken up by mutations where the non-reference count is <span class="math notranslate nohighlight">\(\leq x\)</span>:</p>
<dl class="function">
<dt id="libsequence.nslx">
<code class="descclassname">libsequence.</code><code class="descname">nslx</code><span class="sig-paren">(</span><em>m: libsequence._libsequence.VariantMatrix</em>, <em>refstate: int</em>, <em>x: int</em><span class="sig-paren">)</span> &#x2192; libsequence._libsequence.VecnSLResults<a class="headerlink" href="#libsequence.nslx" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<div class="highlight-ipython notranslate"><div class="highlight"><pre><span></span><span class="gp">In [16]: </span><span class="n">nslx</span> <span class="o">=</span> <span class="n">libsequence</span><span class="o">.</span><span class="n">nslx</span><span class="p">(</span><span class="n">vm</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
</pre></div>
</div>
<p>The haplotype diversity statistics from <a class="reference internal" href="zreferences.html#garud2015-ob" id="id11">[4]</a>:</p>
<dl class="function">
<dt id="libsequence.garud_statistics">
<code class="descclassname">libsequence.</code><code class="descname">garud_statistics</code><span class="sig-paren">(</span><em>arg0: libsequence._libsequence.VariantMatrix</em><span class="sig-paren">)</span> &#x2192; libsequence._libsequence.GarudStats<a class="headerlink" href="#libsequence.garud_statistics" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="class">
<dt id="libsequence.GarudStats">
<em class="property">class </em><code class="descclassname">libsequence.</code><code class="descname">GarudStats</code><a class="headerlink" href="#libsequence.GarudStats" title="Permalink to this definition">¶</a></dt>
<dd><dl class="attribute">
<dt id="libsequence.GarudStats.H1">
<code class="descname">H1</code><a class="headerlink" href="#libsequence.GarudStats.H1" title="Permalink to this definition">¶</a></dt>
<dd><p>Value of H1</p>
</dd></dl>

<dl class="attribute">
<dt id="libsequence.GarudStats.H12">
<code class="descname">H12</code><a class="headerlink" href="#libsequence.GarudStats.H12" title="Permalink to this definition">¶</a></dt>
<dd><p>Value of H2</p>
</dd></dl>

<dl class="attribute">
<dt id="libsequence.GarudStats.H2H1">
<code class="descname">H2H1</code><a class="headerlink" href="#libsequence.GarudStats.H2H1" title="Permalink to this definition">¶</a></dt>
<dd><p>Value of H2/H1</p>
</dd></dl>

</dd></dl>

<div class="highlight-ipython notranslate"><div class="highlight"><pre><span></span><span class="gp">In [17]: </span><span class="n">g</span> <span class="o">=</span> <span class="n">libsequence</span><span class="o">.</span><span class="n">garud_statistics</span><span class="p">(</span><span class="n">vm</span><span class="p">)</span>

<span class="gp">In [18]: </span><span class="k">print</span><span class="p">(</span><span class="n">g</span><span class="o">.</span><span class="n">H1</span><span class="p">,</span> <span class="n">g</span><span class="o">.</span><span class="n">H12</span><span class="p">,</span> <span class="n">g</span><span class="o">.</span><span class="n">H2H1</span><span class="p">)</span>
<span class="go">0.00020202020202020332 0.0006060606060606074 6.4401609045638545e-15</span>
</pre></div>
</div>
<dl class="function">
<dt id="libsequence.two_locus_haplotype_counts">
<code class="descclassname">libsequence.</code><code class="descname">two_locus_haplotype_counts</code><span class="sig-paren">(</span><em>arg0: libsequence._libsequence.VariantMatrix</em>, <em>arg1: int</em>, <em>arg2: int</em>, <em>arg3: bool</em><span class="sig-paren">)</span> &#x2192; List[Sequence::TwoLocusCounts]<a class="headerlink" href="#libsequence.two_locus_haplotype_counts" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="function">
<dt id="libsequence.allele_counts">
<code class="descclassname">libsequence.</code><code class="descname">allele_counts</code><span class="sig-paren">(</span><em>arg0: libsequence._libsequence.AlleleCountMatrix</em><span class="sig-paren">)</span> &#x2192; List[libsequence._libsequence.AlleleCounts]<a class="headerlink" href="#libsequence.allele_counts" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="function">
<dt id="libsequence.non_reference_allele_counts">
<code class="descclassname">libsequence.</code><code class="descname">non_reference_allele_counts</code><span class="sig-paren">(</span><em>*args</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#libsequence.non_reference_allele_counts" title="Permalink to this definition">¶</a></dt>
<dd><p>Overloaded function.</p>
<ol class="arabic simple">
<li>non_reference_allele_counts(arg0: libsequence._libsequence.AlleleCountMatrix, arg1: int) -&gt; List[libsequence._libsequence.AlleleCounts]</li>
<li>non_reference_allele_counts(arg0: libsequence._libsequence.AlleleCountMatrix, arg1: List[int]) -&gt; List[libsequence._libsequence.AlleleCounts]</li>
</ol>
</dd></dl>

<dl class="class">
<dt id="libsequence.AlleleCounts">
<em class="property">class </em><code class="descclassname">libsequence.</code><code class="descname">AlleleCounts</code><a class="headerlink" href="#libsequence.AlleleCounts" title="Permalink to this definition">¶</a></dt>
<dd><dl class="attribute">
<dt id="libsequence.AlleleCounts.nmissing">
<code class="descname">nmissing</code><a class="headerlink" href="#libsequence.AlleleCounts.nmissing" title="Permalink to this definition">¶</a></dt>
<dd><p>Number of samples with missing states</p>
</dd></dl>

<dl class="attribute">
<dt id="libsequence.AlleleCounts.nstates">
<code class="descname">nstates</code><a class="headerlink" href="#libsequence.AlleleCounts.nstates" title="Permalink to this definition">¶</a></dt>
<dd><p>Number of samples with non-missing states</p>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="distribution-of-tajima-s-d-from-msprime">
<h2>Distribution of Tajima’s D from msprime<a class="headerlink" href="#distribution-of-tajima-s-d-from-msprime" title="Permalink to this headline">¶</a></h2>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">msprime</span>
<span class="kn">import</span> <span class="nn">libsequence</span>
<span class="kn">import</span> <span class="nn">seaborn</span> <span class="kn">as</span> <span class="nn">sns</span>

<span class="n">D</span> <span class="o">=</span> <span class="p">[]</span>

<span class="k">for</span> <span class="n">ts</span> <span class="ow">in</span> <span class="n">msprime</span><span class="o">.</span><span class="n">simulate</span><span class="p">(</span><span class="mi">100</span><span class="p">,</span> <span class="n">num_replicates</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span>
                           <span class="n">mutation_rate</span><span class="o">=</span><span class="mf">25.0</span><span class="p">,</span>
                           <span class="n">random_seed</span><span class="o">=</span><span class="mi">42</span><span class="p">):</span>
    <span class="n">m</span> <span class="o">=</span> <span class="n">libsequence</span><span class="o">.</span><span class="n">VariantMatrix</span><span class="o">.</span><span class="n">from_TreeSequence</span><span class="p">(</span><span class="n">ts</span><span class="p">)</span>
    <span class="n">ac</span> <span class="o">=</span> <span class="n">m</span><span class="o">.</span><span class="n">count_alleles</span><span class="p">()</span>
    <span class="n">D</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">libsequence</span><span class="o">.</span><span class="n">tajd</span><span class="p">(</span><span class="n">ac</span><span class="p">))</span>

<span class="n">sns</span><span class="o">.</span><span class="n">distplot</span><span class="p">(</span><span class="n">D</span><span class="p">)</span>
</pre></div>
</div>
<p>(<a class="reference external" href="../pyplots/tajd.py">Source code</a>, <a class="reference external" href="../pyplots/tajd.png">png</a>, <a class="reference external" href="../pyplots/tajd.hires.png">hires.png</a>, <a class="reference external" href="../pyplots/tajd.pdf">pdf</a>)</p>
<div class="figure">
<img alt="../_images/tajd.png" src="../_images/tajd.png" />
</div>
</div>
<div class="section" id="sliding-window-analysis">
<h2>“Sliding window” analysis<a class="headerlink" href="#sliding-window-analysis" title="Permalink to this headline">¶</a></h2>
<p>To analyze different genomic intervals, or “windows”, you have two options.
Any statistic needing the allele count data can easily be gotten via a slice
of your existing data:</p>
<div class="highlight-ipython notranslate"><div class="highlight"><pre><span></span><span class="gp">In [19]: </span><span class="n">lefts</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mf">0.2</span><span class="p">)</span>

<span class="gp">In [20]: </span><span class="k">for</span> <span class="n">l</span> <span class="ow">in</span> <span class="n">lefts</span><span class="p">:</span>
<span class="gp">   ....: </span>    <span class="n">p</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">((</span><span class="n">vm</span><span class="o">.</span><span class="n">positions</span> <span class="o">&gt;=</span> <span class="n">l</span><span class="p">)</span><span class="o">&amp;</span><span class="p">(</span><span class="n">vm</span><span class="o">.</span><span class="n">positions</span> <span class="o">&lt;</span> <span class="n">l</span><span class="o">+</span><span class="mf">0.2</span><span class="p">))[</span><span class="mi">0</span><span class="p">]</span>
<span class="gp">   ....: </span>    <span class="n">ac_slice</span><span class="o">=</span><span class="n">ac</span><span class="p">[</span><span class="n">p</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span><span class="n">p</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">+</span><span class="mi">1</span><span class="p">]</span>
<span class="gp">   ....: </span>    <span class="k">assert</span> <span class="nb">len</span><span class="p">(</span><span class="n">ac_slice</span><span class="p">)</span> <span class="o">==</span> <span class="nb">len</span><span class="p">(</span><span class="n">p</span><span class="p">)</span>
<span class="gp">   ....: </span>    <span class="k">print</span><span class="p">(</span><span class="n">libsequence</span><span class="o">.</span><span class="n">thetapi</span><span class="p">(</span><span class="n">ac_slice</span><span class="p">))</span>
<span class="gp">   ....: </span>
<span class="go">179.54282828282925</span>
<span class="go">198.07858585858628</span>
<span class="go">230.90202020202207</span>
<span class="go">196.7565656565662</span>
<span class="go">175.0989898989905</span>
</pre></div>
</div>
<p>We may also use <a class="reference internal" href="../index.html#libsequence.VariantMatrix.window" title="libsequence.VariantMatrix.window"><code class="xref py py-func docutils literal notranslate"><span class="pre">libsequence.VariantMatrix.window()</span></code></a> to obtain new <cite>VariantMatrix</cite> instances
corresponding to sub-windows of the data:</p>
<div class="highlight-ipython notranslate"><div class="highlight"><pre><span></span><span class="gp">In [21]: </span><span class="k">for</span> <span class="n">l</span> <span class="ow">in</span> <span class="n">lefts</span><span class="p">:</span>
<span class="gp">   ....: </span>    <span class="n">w</span> <span class="o">=</span> <span class="n">vm</span><span class="o">.</span><span class="n">window</span><span class="p">(</span><span class="n">l</span><span class="p">,</span> <span class="n">l</span><span class="o">+</span><span class="mf">0.2</span><span class="p">)</span>
<span class="gp">   ....: </span>    <span class="n">acw</span> <span class="o">=</span> <span class="n">w</span><span class="o">.</span><span class="n">count_alleles</span><span class="p">()</span>
<span class="gp">   ....: </span>    <span class="k">print</span><span class="p">(</span><span class="n">libsequence</span><span class="o">.</span><span class="n">thetapi</span><span class="p">(</span><span class="n">acw</span><span class="p">))</span>
<span class="gp">   ....: </span>
<span class="go">179.54282828282925</span>
<span class="go">198.07858585858628</span>
<span class="go">230.90202020202207</span>
<span class="go">196.7565656565662</span>
<span class="go">175.0989898989905</span>
</pre></div>
</div>
<p>Window creation is <span class="math notranslate nohighlight">\(O(log(vm.nsites))\)</span> in time and has trivial additional memory requirements,
as the returned object does not own its own data buffer.</p>
</div>
<div class="section" id="other-useful-statistics">
<h2>Other useful statistics<a class="headerlink" href="#other-useful-statistics" title="Permalink to this headline">¶</a></h2>
<p>Some more complex descriptors of the data are available</p>
<div class="highlight-ipython notranslate"><div class="highlight"><pre><span></span><span class="gp">In [22]: </span><span class="n">diffs</span> <span class="o">=</span> <span class="n">libsequence</span><span class="o">.</span><span class="n">difference_matrix</span><span class="p">(</span><span class="n">vm</span><span class="p">)</span>
</pre></div>
</div>
<p>Here, <cite>diffs</cite> contains data for representing the distance matrix for the data.  Specifically, these values can be used
to fill the upper triangle of a matrix:</p>
<div class="highlight-ipython notranslate"><div class="highlight"><pre><span></span><span class="gp">In [23]: </span><span class="n">dm</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">diffs</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">int32</span><span class="p">)</span>

<span class="gp">In [24]: </span><span class="n">m</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="mi">0</span><span class="p">]</span><span class="o">*</span><span class="p">(</span><span class="n">vm</span><span class="o">.</span><span class="n">nsam</span><span class="o">*</span><span class="n">vm</span><span class="o">.</span><span class="n">nsam</span><span class="p">))</span>

<span class="go"># Get the indices of the upper-diagonal</span>
<span class="go"># of an nsam*nsam matrix, ignoring</span>
<span class="go"># the diagonal:</span>
<span class="gp">In [25]: </span><span class="n">idx</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">triu_indices</span><span class="p">(</span><span class="n">vm</span><span class="o">.</span><span class="n">nsam</span><span class="p">,</span><span class="n">k</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>

<span class="gp">In [26]: </span><span class="n">m</span><span class="o">=</span><span class="n">m</span><span class="o">.</span><span class="n">reshape</span><span class="p">((</span><span class="n">vm</span><span class="o">.</span><span class="n">nsam</span><span class="p">,</span><span class="n">vm</span><span class="o">.</span><span class="n">nsam</span><span class="p">))</span>

<span class="gp">In [27]: </span><span class="n">m</span><span class="p">[</span><span class="n">idx</span><span class="p">]</span><span class="o">=</span><span class="n">dm</span>

<span class="gp">In [28]: </span><span class="k">print</span><span class="p">(</span><span class="n">m</span><span class="p">)</span>
<span class="go">[[   0 1053  940 ...  940  901 1017]</span>
<span class="go"> [   0    0 1037 ... 1023  898 1038]</span>
<span class="go"> [   0    0    0 ... 1102 1031 1101]</span>
<span class="go"> ...</span>
<span class="go"> [   0    0    0 ...    0  853  893]</span>
<span class="go"> [   0    0    0 ...    0    0 1074]</span>
<span class="go"> [   0    0    0 ...    0    0    0]]</span>

<span class="go"># Let&#39;s confirm our results via brute-force:</span>
<span class="gp">In [29]: </span><span class="n">gm</span> <span class="o">=</span> <span class="n">ts</span><span class="o">.</span><span class="n">genotype_matrix</span><span class="p">()</span>

<span class="gp">In [30]: </span><span class="n">dummy</span> <span class="o">=</span> <span class="mi">0</span>

<span class="gp">In [31]: </span><span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">gm</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">-</span><span class="mi">1</span><span class="p">):</span>
<span class="gp">   ....: </span>    <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">i</span><span class="o">+</span><span class="mi">1</span><span class="p">,</span><span class="n">gm</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]):</span>
<span class="gp">   ....: </span>        <span class="n">diffs</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">gm</span><span class="p">[:,</span><span class="n">i</span><span class="p">]</span> <span class="o">!=</span> <span class="n">gm</span><span class="p">[:,</span><span class="n">j</span><span class="p">])</span>
<span class="gp">   ....: </span>        <span class="k">assert</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">diffs</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span><span class="o">==</span><span class="n">dm</span><span class="p">[</span><span class="n">dummy</span><span class="p">])</span>
<span class="gp">   ....: </span>        <span class="n">dummy</span><span class="o">+=</span><span class="mi">1</span>
<span class="gp">   ....: </span>
</pre></div>
</div>
<p>In a similar fashion, we can obtain true/false data on whether pairs of haplotypes differ:</p>
<div class="highlight-ipython notranslate"><div class="highlight"><pre><span></span><span class="gp">In [32]: </span><span class="n">diff_yes_or_no</span> <span class="o">=</span> <span class="n">libsequence</span><span class="o">.</span><span class="n">is_different_matrix</span><span class="p">(</span><span class="n">vm</span><span class="p">)</span>
</pre></div>
</div>
<dl class="function">
<dt id="libsequence.is_different_matrix">
<code class="descclassname">libsequence.</code><code class="descname">is_different_matrix</code><span class="sig-paren">(</span><em>m: libsequence._libsequence.VariantMatrix</em><span class="sig-paren">)</span> &#x2192; List[int]<a class="headerlink" href="#libsequence.is_different_matrix" title="Permalink to this definition">¶</a></dt>
<dd><p>Return whether or not pairs of 
samples in a VariantMatrix differ</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>m</strong> – A <a class="reference internal" href="../index.html#libsequence.VariantMatrix" title="libsequence.VariantMatrix"><code class="xref py py-class docutils literal notranslate"><span class="pre">libsequence.VariantMatrix</span></code></a></td>
</tr>
</tbody>
</table>
</dd></dl>

<p>The contents of this matrix have the exact same layout as <cite>diffs</cite> described above.  The difference is that the data
elements are encoded as 0 = identical, 1 = different.  This calculation is <strong>much</strong> faster than the previous.</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">Missing data do not contribute to samples being considered different, nor to the
number of differences.</p>
</div>
<p>It is also possible to get a unique label assigned to each haplotype:</p>
<div class="highlight-ipython notranslate"><div class="highlight"><pre><span></span><span class="gp">In [33]: </span><span class="n">labels</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">libsequence</span><span class="o">.</span><span class="n">label_haplotypes</span><span class="p">(</span><span class="n">vm</span><span class="p">),</span><span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">int32</span><span class="p">)</span>

<span class="gp">In [34]: </span><span class="k">print</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">unique</span><span class="p">(</span><span class="n">labels</span><span class="p">)))</span>
<span class="go">99</span>
</pre></div>
</div>
<dl class="function">
<dt id="libsequence.label_haplotypes">
<code class="descclassname">libsequence.</code><code class="descname">label_haplotypes</code><span class="sig-paren">(</span><em>arg0: libsequence._libsequence.VariantMatrix</em><span class="sig-paren">)</span> &#x2192; List[int]<a class="headerlink" href="#libsequence.label_haplotypes" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<p>Internally, the results from <cite>is_different_matrix</cite> are used to assign the labels.</p>
<p>These labels are likewise used internally to count the number of haplotypes:</p>
<div class="highlight-ipython notranslate"><div class="highlight"><pre><span></span><span class="gp">In [35]: </span><span class="k">print</span><span class="p">(</span><span class="n">libsequence</span><span class="o">.</span><span class="n">number_of_haplotypes</span><span class="p">(</span><span class="n">vm</span><span class="p">))</span>
<span class="go">99</span>

<span class="go"># Confirm result via direct comparison to</span>
<span class="go"># the data from msprime:</span>
<span class="gp">In [36]: </span><span class="k">print</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">unique</span><span class="p">(</span><span class="n">gm</span><span class="o">.</span><span class="n">transpose</span><span class="p">(),</span><span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)))</span>
<span class="go">   99</span>
</pre></div>
</div>
<p>What about performance?</p>
<div class="highlight-ipython notranslate"><div class="highlight"><pre><span></span><span class="gp">In [37]: </span><span class="o">%</span><span class="k">timeit</span> -n 10 -r 10 libsequence.number_of_haplotypes(vm)
<span class="go">10 loops, best of 10: 272 us per loop</span>
</pre></div>
</div>
<div class="highlight-ipython notranslate"><div class="highlight"><pre><span></span><span class="gp">In [38]: </span><span class="o">%</span><span class="k">timeit</span> -n 10 -r 10 len(np.unique(gm.transpose(),axis=0))
<span class="go">10 loops, best of 10: 11.6 ms per loop</span>
</pre></div>
</div>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="../index.html">Table of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Calculating summary statistics</a><ul>
<li><a class="reference internal" href="#simple-overview-of-concepts">Simple overview of concepts</a></li>
<li><a class="reference internal" href="#list-of-classic-standard-summary-statistics">List of “classic”/standard summary statistics</a></li>
<li><a class="reference internal" href="#distribution-of-tajima-s-d-from-msprime">Distribution of Tajima’s D from msprime</a></li>
<li><a class="reference internal" href="#sliding-window-analysis">“Sliding window” analysis</a></li>
<li><a class="reference internal" href="#other-useful-statistics">Other useful statistics</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="vmatrix.html"
                        title="previous chapter">The VariantMatrix</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="treesequences.html"
                        title="next chapter">Tree Sequences</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="../_sources/pages/summstats.rst.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    </div>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="treesequences.html" title="Tree Sequences"
             >next</a> |</li>
        <li class="right" >
          <a href="vmatrix.html" title="The VariantMatrix"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="../index.html">pylibseq 0.2.2 documentation</a> &#187;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &#169; Copyright 2015, Kevin Thornton.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.8.5.
    </div>
  </body>
</html>